### PR TITLE
Sync UMA: fairchem UMA Core with shared Q/S readouts

### PR DESCRIPTION
**Pull request by Benzoin96485**
## Summary
- Merge `UMA` onto `sync/datahub-devel`: `uma_qs` Core (`UMAWrapperQS`), shared `SimpleReadout` / spin-conservation layers
- Reconcile `datahub.py`: keep ASE energy/force unit conversion **and** Q/S from `atoms.info`
- Docs (architecture catalog, layer stack, README) + lightweight registry tests (no fairchem load)

## Test plan
- [x] `pytest test/test_aselmdb.py test/test_uma_registry.py`
- [x] Smoke train `architecture: uma_qs` with a fairchem checkpoint (GPU / large download)

Stacked PR #2 of 3. Base is `sync/datahub-devel` (#80); retarget to `devel` after #80 merges.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by Benzoin96485** [comment]
bugbot review

File: enerzyme/tasks/batch.py [hunk 1]
```diff
@@ -92,20 +93,24 @@ def _decorate_batch_input(batch: Iterable[Tuple[Dict[str, Tensor], Dict[str, Ten
...
         if "idx_i" in feature:
             batch_idx_i.append(feature["idx_i"][:feature["N_pair"]] + count)
             batch_idx_j.append(feature["idx_j"][:feature["N_pair"]] + count)
-        else:
+            built_graph = True
+        elif otf_graph and not built_graph:
             idx_i, idx_j = full_neighbor_list(feature["N"])
             batch_idx_i.append(idx_i + count)
             batch_idx_j.append(idx_j + count)
+            built_graph = True
```
[the remark this answers is not in the document]
Comment (Benzoin96485) [reply enerzyme/tasks/batch.py]: Fixed in a666514.

Confirmed: the batch-wide `built_graph` flag stopped OTF neighbor construction after the first structure. `_decorate_batch_input` now builds `idx_i`/`idx_j` per molecule (precomputed or OTF), and raises if a batch would otherwise concatenate a partial edge list.

Coverage in `test/test_uma_qs_train_smoke.py` (OTF multi-molecule + mixed precomputed/OTF + incomplete-list rejection). Smoke config uses `batch_size: 2` with `neighbor_list: ""`; live train collates two fixture frames and asserts a full per-molecule graph before the epoch. Verified on GPU with `enerzyme-dev`: `pytest test/test_uma_qs_train_smoke.py` → 5 passed.

[the remark this answers is not in the document]
Comment (Benzoin96485) [reply enerzyme/tasks/batch.py]: Follow-up (clarification + broader coverage) in f9eaf20.

**Impact on `uma_qs`:** `UMAWrapperQS` does **not** use Enerzyme `idx_i` / `idx_j`. Its `input_fields` are `Ra` / `Za` / `batch_seg` / `Q` / `S`; it builds fairchem `AtomicData` with an empty `edge_index` and sets `backbone.otf_graph = True`, so the neighbor graph is constructed inside fairchem. The incomplete Enerzyme edge list from the old `built_graph` flag would **not** change `uma_qs` predictions. The earlier uma smoke graph asserts check the collate contract, not a UMA forward dependency.

**Where the bug *does* matter:** SpookyNet / PhysNet / SchNet take `idx_i_sr` / `idx_j_sr` from `RangeSeparation` on Enerzyme edges. With `neighbor_list: ""` and `batch_size > 1`, the prior flag would silently drop edges for later molecules in the batch.

**New always-on coverage** (`test/test_otf_batch_arch_smoke.py`, no large checkpoint):
- Fixture `molecules_otf_tiny.pkl` (H₂ / H₂O / CH₄, N=2/3/5)
- Configs `train_otf_batch_{spookynet,physnet,schnet}.yaml` with `neighbor_list: ""`, `batch_size: 2`, shrunk cores
- Per architecture: OTF collate full-graph assert → forward finite `E`/`Fa` → 1-epoch `FFTrain` + `model*.pth`

Verified on GPU (`enerzyme-dev`): `pytest test/test_otf_batch_arch_smoke.py` → **9 passed**.

File: enerzyme/models/layers/readout.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        self.heads = ModuleList(self._get_head())
+        self.use_nhloss = use_nhloss
+
+    def get_output(self, atom_feature: Tensor) -> Dict[str, Tensor]:
+        raw_output = 0.
+        if self.use_nhloss:
+            nhloss = 0.
+            lastoutput2 = 0.
+        for i in range(self.num_blocks):
+            raw_output += self.heads[i](atom_feature[:, :, i])
```
[the remark this answers is not in the document]
Comment (Benzoin96485) [reply enerzyme/models/layers/readout.py:164]: Fixed in 8a2c4fb.

Confirmed: `ModuleList(self._get_head())` wrapped a single head (current PyTorch raises `TypeError`; older builds could end up empty), and `BaseReadout` never assigned `self.num_blocks`. Now `self.num_blocks` is stored and heads are built as `ModuleList([self._get_head() for _ in range(self.num_blocks)])`. Verified forward with 3D `(natoms, dim, num_blocks)` features for dense / residual_layer heads (including `nh_loss`).

File: enerzyme/tasks/batch.py [hunk 3]
```diff
@@ -27,11 +27,12 @@ def _decorate_pyg_batch_input(batch: Iterable[Tuple[Dict[str, Tensor], Dict[str,
...
             data_dict["idx_i"] = torch.tensor(feature["idx_i"], dtype=torch.long)
             data_dict["idx_j"] = torch.tensor(feature["idx_j"], dtype=torch.long)
-        else:
+            feature_list.append(Data(edge_index=torch.stack([data_dict["idx_i"], data_dict["idx_j"]], dim=0), num_nodes=feature["N"], **data_dict))
+        elif otf_graph:
             idx_i, idx_j = full_neighbor_list(feature["N"])
             data_dict["idx_i"] = torch.tensor(idx_i, dtype=torch.long)
             data_dict["idx_j"] = torch.tensor(idx_j, dtype=torch.long)
-        feature_list.append(Data(edge_index=torch.stack([data_dict["idx_i"], data_dict["idx_j"]], dim=0), num_nodes=feature["N"], **data_dict))
+            feature_list.append(Data(edge_index=torch.stack([data_dict["idx_i"], data_dict["idx_j"]], dim=0), num_nodes=feature["N"], **data_dict))
```
[the remark this answers is not in the document]
Comment (Benzoin96485) [reply enerzyme/tasks/batch.py]: Fixed in e832286.

Confirmed: with `otf_graph=False` and no precomputed `idx_i`/`idx_j`, `_decorate_pyg_batch_input` never appended to `feature_list`, so `Batch.from_data_list([])` raised `IndexError` while targets were still built for every sample.

Now every sample is always appended (empty `edge_index` when edges are omitted, for UMA-style cores that build the graph themselves). Mixed batches (some with edges, some without) raise the same incomplete-neighbor-list `ValueError` as the non-PyG path.

Regression coverage in `test/test_uma_qs_train_smoke.py`:
- `test_decorate_pyg_batch_input_keeps_samples_without_edges`
- `test_decorate_pyg_batch_input_rejects_incomplete_neighbor_lists`

Verified with `enerzyme-dev`: both new tests + incomplete-list rejection → 3 passed.

**Comment by Benzoin96485** [comment]
bugbot review

File: enerzyme/models/layers/readout.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        self.head = self._get_head()
+
+    def get_output(self, atom_feature: Tensor) -> Dict[str, Tensor]:
+        if atom_feature.ndim == 2:
+            output = self.head(atom_feature)
+        elif atom_feature.ndim == 3:
+            output = self.head(atom_feature[:, :, -1])
+        return {
+            self.ordered_output_fields[i]: output[:, i] for i in range(self.dim_feature_out)
+        }
```
[the remark this answers is not in the document]
Comment (Benzoin96485) [reply enerzyme/models/layers/readout.py]: False positive — no runtime bug here.

`DenseLayer` with `shallow_ensemble_size > 1` intentionally returns `(N, fields, ensemble)`. Selecting field `i` yields `(N, ensemble)`, matching PhysNet / SchNet / MACE and the contract expected by ChargeConservation, AtomicAffine, Force, `WeightedLoss`, and `ShallowEnsembleReduce`. Collapsing to per-atom scalars would break shallow ensembles.

Clarified via `_split_field_outputs` (shape checks + docs) and locked with `test/test_readout_shallow_ensemble.py` (7/7 passed on `d-7-5-1` / `enerzyme-dev`) in e6ec40a.
